Alg

obroadrick opened the pull request:
Feedback?

Items to discuss:
ensure that inputs are valid (types, sizes, etc)
ensure sorted tasks (either here or there)
ensure compatable return format
stablish how to handle an unsolvable input set (ie. no viable schedule exists)

obroadrick committed:
Implemented scheduling API and internal algorithm

- based on scheduling algorithm by Yao et al
- added function to compute reward from precision and priority
- added test file that initially just runs the algorithm on a toy example

obroadrick committed:
remove notes

obroadrick committed:
add gitignore

grahamschock reviewed (changes requested):
I think these are some starting things to talk about. Can add more comments later.

obroadrick committed:
make an Algorithm class to wrap the functions in alg.py

obroadrick committed:
class modified according to PR feedback

obroadrick committed:
alg functions input tests/checks

obroadrick committed:
add simulation script

obroadrick committed:
remove pycache

obroadrick committed:
sim script

obroadrick committed:
few scripts, will consolidate based on new direction

obroadrick committed:
fix depth allocation and reward range

when 0 reward remains when building the depth schedule, the algorithm now
correclty allocates 0 (as depth, reward, and time) to the remaining
tasks (so as to avoid indexing into "None" solutions (which are correct)
in the P table)

fix the reward range of tables to be rmax_quantized + 1 (the plus one bit
being the new bit) since this allows for 0 reward up to the max reward.

obroadrick committed:
Change the S and P tables to have correct length

they had the maximum possible reward for a single task, but of course
they should have that value times the number of tasks

obroadrick committed:
simplify to a single simulation script

obroadrick committed:
make delta a member of alg class

obroadrick committed:
change sampling choices as per discussion with team

obroadrick committed:
add basic metrics for measuring alg performance

basic function that checks validity of schedule (finish w/in deadline, etc)

basic metric that computes the achieved reward=prec*prio

obroadrick committed:
add metrics, plotting of sim results

grahamschock commented on alg.py:
```diff
@@ -1,182 +1,367 @@
+            if depth_sched == None:
+                print("No viable schedule for these inputs")
+                return None
+            print("Depth schedule that achieves reward", sum(reward_sched), "  ")
```
We shouldn't be printing here if this API is going to be used for an actual scheduler

obroadrick replied:
There are many print statements throughout that provide useful information from our research perspective, but I have kept all of them in "if verbose" blocks. So, when called from the scheduler, we can simply use "verbose=False" to avoid them, but this way we can keep them around for the research side of things... after all we aren't making production code, we are doing research

grahamschock replied:
listen pal... there wouldn't be any "research" if there we didn't have production code. When we finally do experiments the we will be looking at the speed of the algorithm with GPU workloads it will be a lot slower if we are printing every quantum.

obroadrick replied:
Ha, the code does not print anything as long as you call with "verbose=False"

grahamschock reviewed (changes requested):
API looks good! I can't really comment on if the algorithm is correct or not but I assume it is. Just change the prints and we should be good to merge with GPU stuff over winter break.
